File: app/src/main/java/com/mxt/anitrend/base/custom/view/text/RichMarkdownTextView.java
package com.mxt.anitrend.base.custom.view.text;

import android.content.Context;
import android.databinding.BindingAdapter;
import android.support.annotation.StringRes;
import android.support.v4.text.util.LinkifyCompat;
import android.support.v7.widget.AppCompatTextView;
import android.text.Html;
import android.text.Spanned;
import android.text.method.LinkMovementMethod;
import android.text.util.Linkify;
import android.util.AttributeSet;
import android.widget.TextView;
import android.widget.Toast;

import com.mxt.anitrend.base.interfaces.view.CustomView;
import com.mxt.anitrend.util.MarkDownUtil;
import com.mxt.anitrend.util.RegexUtil;

import ru.noties.markwon.Markwon;
import ru.noties.markwon.SpannableConfiguration;
import ru.noties.markwon.il.AsyncDrawableLoader;

public class RichMarkdownTextView extends AppCompatTextView implements CustomView {

    public RichMarkdownTextView(Context context) {
        super(context);
        onInit();
    }

    public RichMarkdownTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
        onInit();
    }

    public RichMarkdownTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        onInit();
    }

    /**
     * Optionally included when constructing custom views
     */
    @Override
    public void onInit() {
        setFocusable(false);
        LinkifyCompat.addLinks(this, Linkify.WEB_URLS);
        setMovementMethod(LinkMovementMethod.getInstance());
    }

    /**
     * Clean up any resources that won't be needed
     */
    @Override
    public void onViewRecycled() {

    }

    @BindingAdapter("markDown")
    public static void markDown(RichMarkdownTextView richMarkdownTextView, String markdown) {
        String strippedText = RegexUtil.removeTags(markdown);
        Spanned markdownSpan = MarkDownUtil.convert(strippedText, richMarkdownTextView.getContext(), richMarkdownTextView);
        richMarkdownTextView.setText(markdownSpan, TextView.BufferType.SPANNABLE);
    }

    @BindingAdapter("textHtml")
    public static void htmlText(RichMarkdownTextView richMarkdownTextView, String html) {
        Spanned markdownSpan = MarkDownUtil.convert(html, richMarkdownTextView.getContext(), richMarkdownTextView);
        richMarkdownTextView.setText(markdownSpan, TextView.BufferType.SPANNABLE);
    }

    @BindingAdapter("basicHtml")
    public static void basicText(RichMarkdownTextView richMarkdownTextView, String html) {
        Spanned htmlSpan = Html.fromHtml(html);
        richMarkdownTextView.setText(htmlSpan);
    }

    @BindingAdapter("textHtml")
    public static void htmlText(RichMarkdownTextView richMarkdownTextView, @StringRes int resId) {
        String text = richMarkdownTextView.getContext().getString(resId);
        Spanned markdownSpan = MarkDownUtil.convert(text, richMarkdownTextView.getContext(), richMarkdownTextView);
        richMarkdownTextView.setText(markdownSpan, TextView.BufferType.SPANNABLE);
    }

    @BindingAdapter("richMarkDown")
    public static void richMarkDown(RichMarkdownTextView richMarkdownTextView, String markdown) {
        final Context context = richMarkdownTextView.getContext();

        final SpannableConfiguration configuration = SpannableConfiguration.builder(context)
                .asyncDrawableLoader(AsyncDrawableLoader.create())
                .htmlAllowNonClosedTags(false)
                .softBreakAddsNewLine(false)
                .build();

        Markwon.setMarkdown(richMarkdownTextView, configuration, RegexUtil.convertToStandardMarkdown(markdown));
    }

    public void setMarkDownText(String markDownText) {
        String strippedText = RegexUtil.removeTags(markDownText);
        Spanned markdownSpan = MarkDownUtil.convert(strippedText, getContext(), this);
        setText(markdownSpan, TextView.BufferType.SPANNABLE);
    }
}

File: app/src/main/java/com/mxt/anitrend/base/custom/view/text/RichMarkdownTextView.kt
package com.mxt.anitrend.base.custom.view.text

import android.content.Context
import android.text.util.Linkify
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatTextView
import com.google.android.material.textview.MaterialTextView
import com.mxt.anitrend.base.interfaces.view.CustomView
import com.mxt.anitrend.binding.richMarkDown
import com.mxt.anitrend.util.DialogUtil
import com.mxt.anitrend.util.Settings
import com.mxt.anitrend.util.markdown.MarkDownUtil
import com.mxt.anitrend.util.markdown.RegexUtil
import com.mxt.anitrend.view.sheet.BottomSheetMessage
import io.noties.markwon.Markwon
import me.saket.bettermovementmethod.BetterLinkMovementMethod
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import org.koin.core.parameter.parametersOf

class RichMarkdownTextView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : MaterialTextView(context, attrs, defStyleAttr), CustomView, KoinComponent {

    val markwon by inject<Markwon>()
    val settings by inject<Settings>()

    init { onInit() }

    /**
     * Optionally included when constructing custom views
     */
    override fun onInit() {
        setTextIsSelectable(true)
        Linkify.addLinks(this, Linkify.WEB_URLS)
        movementMethod = BetterLinkMovementMethod.getInstance()
    }

    /**
     * Clean up any resources that won't be needed
     */
    override fun onViewRecycled() {

    }

    fun setMarkDownText(markDownText: String?) {
        if (!settings.experimentalMarkdown) {
            val strippedText = RegexUtil.removeTags(markDownText)
            val markdownSpan = MarkDownUtil.convert(strippedText)
            setText(markdownSpan, BufferType.SPANNABLE)
        } else markwon.setMarkdown(this, markDownText ?: "**No content available**")
    }
}
